crates/sui-quorum-driver/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use arc_swap::ArcSwap;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time::Instant;
+
+use tokio::sync::mpsc::{self, Receiver, Sender};
+use tokio::sync::Mutex;
+use tokio::task::JoinHandle;
+use tracing::log::{error, warn};
+use tracing::Instrument;
+
+use sui_core::authority_aggregator::AuthorityAggregator;
+use sui_core::authority_client::AuthorityAPI;
+use sui_types::error::{SuiError, SuiResult};
+use sui_types::messages::{
+    CertifiedTransaction, ExecuteTransactionRequest, ExecuteTransactionRequestType,
+    ExecuteTransactionResponse, Transaction, TransactionEffects,
+};
+
+const DEFAULT_PROCESS_TIMEOUT: Duration = Duration::from_secs(30);
+
+enum QuorumTask<A> {
+    ProcessTransaction(Transaction),
+    ProcessCertificate(CertifiedTransaction),
+    UpdateValidators(AuthorityAggregator<A>),
+}
+
+pub struct QuorumDriverHandler<A> {
+    quorum_driver: Arc<QuorumDriver<A>>,
+    _processor_handle: JoinHandle<()>,
+    task_sender: Mutex<Sender<QuorumTask<A>>>,
+    // TODO: Change to CertifiedTransactionEffects eventually.
+    effects_subscriber: Mutex<Receiver<(CertifiedTransaction, TransactionEffects)>>,
```

Comment:
Same - I think no Mutex is necessary

Reply:
Receiver::recv() takes a mut ref self, so it seems I have to use a mutex

Reply:
Could you use a broadcast channel rather than a mutex over a mpsc channel?

Reply:
If only 1 task is meant to receive on a channel, then it should have exclusive ownership over said channel and it should not need to be wrapped in a mutex.